python/dgl/data/csv_dataset.py

```diff
@@ -0,0 +1,420 @@
+        edge_dict = DGLGraphConstructor._parse_edge_data(edge_data, node_dict)
+        graph_dict = DGLGraphConstructor._construct_graphs(
+            node_dict, edge_dict)
+        if graph_data is None:
+            graph_data = GraphData(np.full(1, 0), {})
+        graphs, data = DGLGraphConstructor._parse_graph_data(
+            graph_data, graph_dict)
+        return graphs, data
+
+    @staticmethod
+    def _parse_node_data(node_data):
+        # node_ids could be arbitrary numeric values, namely non-sorted, duplicated, not labeled from 0 to num_nodes-1
+        node_dict = {}
+        for n_data in node_data:
+            graph_ids = np.unique(n_data.graph_id)
+            for graph_id in graph_ids:
+                if graph_id in node_dict and n_data.type in node_dict[graph_id]:
+                    raise DGLError(f"Duplicate node type[{n_data.type}] for same graph[{graph_id}], please place the same node_type for same graph into single NodeData.")
+                idx = n_data.graph_id == graph_id
+                ids = n_data.id[idx]
+                u_ids, u_indices = np.unique(ids, return_index=True)
+                if len(ids) > len(u_ids):
+                    dgl_warning(
+                        "There exist duplicated ids and only the first ones are kept.")
+                ndata = {}
+                ndata['mapping'] = {index: i for i,
+                                    index in enumerate(ids[u_indices])}
+                data = {}
+                for key, value in n_data.data.items():
+                    data[key] = F.tensor(value[idx][u_indices])
+                ndata['data'] = data
+                if graph_id not in node_dict:
+                    node_dict[graph_id] = {}
+                node_dict[graph_id][n_data.type] = ndata
+        return node_dict
+
+    @staticmethod
+    def _parse_edge_data(edge_data, node_dict):
+        edge_dict = {}
+        for e_data in edge_data:
+            (src_type, e_type, dst_type) = e_data.type
+            graph_ids = np.unique(e_data.graph_id)
+            for graph_id in graph_ids:
+                if graph_id in edge_dict and e_data.type in edge_dict[graph_id]:
+                    raise DGLError(f"Duplicate edge type[{e_data.type}] for same graph[{graph_id}], please place the same edge_type for same graph into single EdgeData.")
+                idx = e_data.graph_id == graph_id
+                edata = {}
+                src_mapping = node_dict[graph_id][src_type]['mapping']
+                dst_mapping = node_dict[graph_id][dst_type]['mapping']
+                src_ids = [src_mapping[index] for index in e_data.src[idx]]
+                dst_ids = [dst_mapping[index] for index in e_data.dst[idx]]
+                edata['edges'] = (
+                    F.tensor(src_ids), F.tensor(dst_ids))
+                data = {}
+                for key, value in e_data.data.items():
+                    data[key] = F.tensor(value[idx])
+                edata['data'] = data
```

Comment:
use list comprehension to shorten the code

Reply:
accepted.